storage/namespace.go

```diff
@@ -0,0 +1,270 @@
+package storage
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/clock"
+	"github.com/m3db/m3db/context"
+	"github.com/m3db/m3db/persist"
+	"github.com/m3db/m3db/persist/fs/commitlog"
+	"github.com/m3db/m3db/sharding"
+	"github.com/m3db/m3db/storage/bootstrap"
+	"github.com/m3db/m3db/storage/namespace"
+	"github.com/m3db/m3db/ts"
+	"github.com/m3db/m3db/x/io"
+	"github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/time"
+)
+
+func commitLogWriteNoOp(
+	series commitlog.Series,
+	datapoint ts.Datapoint,
+	unit xtime.Unit,
+	annotation ts.Annotation,
+) error {
+	return nil
+}
+
+type dbNamespace struct {
+	sync.RWMutex
+
+	name     string
+	shardSet sharding.ShardSet
+	nopts    namespace.Options
+	dopts    Options
+	nowFn    clock.NowFn
+	bs       bootstrapState
+
+	// Contains an entry to all shards for fast shard lookup, an
+	// entry will be nil when this shard does not belong to current database
+	shards []databaseShard
+
+	increasingIndex  increasingIndex
+	writeCommitLogFn writeCommitLogFn
+}
+
+func newDatabaseNamespace(
+	metadata namespace.Metadata,
+	shardSet sharding.ShardSet,
+	increasingIndex increasingIndex,
+	writeCommitLogFn writeCommitLogFn,
+	dopts Options,
+) databaseNamespace {
+	name := metadata.Name()
+	nopts := metadata.Options()
+	fn := writeCommitLogFn
+	if !nopts.GetWritesToCommitLog() {
+		fn = commitLogWriteNoOp
+	}
+
+	n := &dbNamespace{
+		name:             name,
+		shardSet:         shardSet,
+		nopts:            nopts,
+		dopts:            dopts,
+		nowFn:            dopts.GetClockOptions().GetNowFn(),
+		increasingIndex:  increasingIndex,
+		writeCommitLogFn: fn,
+	}
+
+	n.initShards()
+
+	return n
+}
+
+func (n *dbNamespace) Name() string {
+	return n.name
+}
+
+func (n *dbNamespace) Tick() {
+	shards := n.getOwnedShards()
+	if len(shards) == 0 {
+		return
+	}
+
+	// Tick through the shards sequentially to avoid parallel data flushes
```

Comment:
I'm also fine if we want to test some of this stuff before adding the splay.  We should profile the spikes more carefully and ensure it's the splay that's the extra hot part of the CPU spikes.